### PR TITLE
Fix static library compiling

### PR DESCRIPTION
**Pull request by madduci**
Tested with VS2015 and Linux/gcc 5.3.0,

the library can be compiled as static/shared and the .so version number is shown at the end of the file.

### CODE REVIEW

File: CMake/Utilities.cmake [hunk 1]
```diff
@@ -29,3 +33,41 @@ function(IN_STR_LIST LIST_NAME ITEM_SEARCHED RETVAL)
...
+        CMAKE_CXX_FLAGS
+        CMAKE_CXX_FLAGS_DEBUG
+        CMAKE_CXX_FLAGS_MINSIZEREL
+        CMAKE_CXX_FLAGS_RELEASE
+        CMAKE_CXX_FLAGS_RELMITDEBINFO)
+
+    #Set /MT flag for Static compiling, /MD for Shared
+    foreach(Flag ${CompilerFlags})
+      if("${BUILD_TYPE}" MATCHES "STATIC")
+        string(REPLACE "/MD" "/MT" ${Flag} "${${CompilerFlags}}")
```
Comment (bagder) [inline CMake/Utilities.cmake]: Surely not all compilers on win32 supports these options? Shouldn't this be made sure to only be used if MSVC is used?

File: CMake/Platforms/WindowsCache.cmake [hunk 2]
```diff
@@ -118,6 +118,11 @@ if(NOT UNIX)
 
     set(HAVE_SIGACTION 0)
     set(HAVE_MACRO_SIGSETJMP 0)
+
+    if(${CMAKE_CL_64})
+      add_definitions(D_AMD64 /Damd64)
```
Comment (bagder) [inline CMake/Platforms/WindowsCache.cmake]: Is this really related to building statically? Is "having a 64 bit compiler from microsoft" really a enough to deduce that we're on an amd64 architecture? (which I assume the define would signify, there's no comment that explains what it's for)

File: CMake/Utilities.cmake [hunk 3]
```diff
@@ -29,3 +33,41 @@ function(IN_STR_LIST LIST_NAME ITEM_SEARCHED RETVAL)
...
+      else()
+        string(REPLACE "/MT" "/MD" ${Flag} "${${CompilerFlags}}")
+      endif()
+    endforeach()
+  else()
+    if("${BUILD_TYPE}" MATCHES "STATIC")
+      set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fPIC")
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
```
Comment (bagder) [inline CMake/Utilities.cmake]: -fPIC unconditionally everywhere, really?

File: CMake/Utilities.cmake [hunk 4]
```diff
@@ -29,3 +33,41 @@ function(IN_STR_LIST LIST_NAME ITEM_SEARCHED RETVAL)
...
+      set(CMAKE_EXE_LINKER_FLAGS "")
+    else()
+      set(CMAKE_EXE_LINKER_FLAGS "-static -static-libgcc -static-libstdc++")
```
Comment (bagder) [inline CMake/Utilities.cmake]: these options seem to assume that you're using gcc? how can you do that here?

File: CMakeLists.txt [hunk 5]
```diff
@@ -1020,9 +1020,9 @@ function(TRANSFORM_MAKEFILE_INC INPUT_FILE OUTPUT_FILE)
   string(REPLACE "$(top_srcdir)"   "\${CURL_SOURCE_DIR}" MAKEFILE_INC_TEXT ${MAKEFILE_INC_TEXT})
   string(REPLACE "$(top_builddir)" "\${CURL_BINARY_DIR}" MAKEFILE_INC_TEXT ${MAKEFILE_INC_TEXT})
 
-  string(REGEX REPLACE "\\\\\n" "�!�" MAKEFILE_INC_TEXT ${MAKEFILE_INC_TEXT})
+  string(REGEX REPLACE "\\\\\n" "�!�" MAKEFILE_INC_TEXT ${MAKEFILE_INC_TEXT})
```
Comment (bagder) [inline CMakeLists.txt]: Was this change really intended? If so, why?

File: CMake/Utilities.cmake (hunk 1 above)
Comment (madduci) [inline CMake/Utilities.cmake]: true, but actually they will be recognized only by Visual Studio

Comment (bagder) [inline CMake/Utilities.cmake]: And Visual Studio is the only windows compiler that can be used with cmake?

File: CMake/Utilities.cmake [hunk 6]
```diff
@@ -29,3 +33,41 @@ function(IN_STR_LIST LIST_NAME ITEM_SEARCHED RETVAL)
...
+        CMAKE_C_FLAGS
+        CMAKE_C_FLAGS_DEBUG
+        CMAKE_C_FLAGS_MINSIZEREL
+        CMAKE_C_FLAGS_RELEASE
+        CMAKE_C_FLAGS_RELMITDEBINFO
+        CMAKE_CXX_FLAGS
+        CMAKE_CXX_FLAGS_DEBUG
+        CMAKE_CXX_FLAGS_MINSIZEREL
+        CMAKE_CXX_FLAGS_RELEASE
+        CMAKE_CXX_FLAGS_RELMITDEBINFO)
```
Comment (Lekensteyn) [inline CMake/Utilities.cmake]: German user? The default configs include "RELWITHDEBINFO".

These lines can be found in `Modules/Platform/Windows-MSVC.cmake` around line 300:
```
  set(CMAKE_${lang}_FLAGS_DEBUG_INIT "/D_DEBUG /MDd /Zi /Ob0 /Od ${_RTC1}")
  set(CMAKE_${lang}_FLAGS_RELEASE_INIT "/MD /O2 /Ob2 /D NDEBUG")
  set(CMAKE_${lang}_FLAGS_RELWITHDEBINFO_INIT "/MD /Zi /O2 /Ob1 /D NDEBUG")
  set(CMAKE_${lang}_FLAGS_MINSIZEREL_INIT "/MD /O1 /Ob1 /D NDEBUG")
```

File: CMakeLists.txt [hunk 7]
```diff
@@ -1020,9 +1031,9 @@ function(TRANSFORM_MAKEFILE_INC INPUT_FILE OUTPUT_FILE)
...
   string(REGEX REPLACE "([a-zA-Z_][a-zA-Z0-9_]*)[\t ]*=[\t ]*([^\n]*)" "SET(\\1 \\2)" MAKEFILE_INC_TEXT ${MAKEFILE_INC_TEXT})
-  string(REPLACE "�!�" "\n" MAKEFILE_INC_TEXT ${MAKEFILE_INC_TEXT})
+  string(REPLACE "�!�" "\n" MAKEFILE_INC_TEXT ${MAKEFILE_INC_TEXT})
```
Comment (Lekensteyn) [inline CMakeLists.txt]: Your editor is broken..

File: CMake/Utilities.cmake (hunk 3 above)
Comment (Lekensteyn) [inline CMake/Utilities.cmake]: Why do you exactly have to set fPIC for static builds? I think you inversed the logic here. All these flags should not be needed if you use `add_library(... STATIC ...)`.

**Comment by madduci** [comment]
i will fix all the editor failures and make stronger checks on the builds.
